Add fast tokenizer for BARTpho

datquocnguyen opened the pull request:
This PR is to add a "fast" BARTpho tokenizer (backed by HuggingFace's *tokenizers* library).

# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

datquocnguyen committed:
Add BartphoTokenizerFast

datquocnguyen committed:
Add BartphoTokenizerFast

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17254). All of your documentation changes will be reflected on that endpoint.

datquocnguyen committed:
Add test for BartphoTokenizerFast

datquocnguyen commented:
Following: [https://github.com/huggingface/transformers/pull/13788](https://github.com/huggingface/transformers/pull/13788)
I now add a "fast" version of the BartphoTokenizer. 
@sgugger , @LysandreJik, @patil-suraj , @SaulLu and @patrickvonplaten Please could you have a look and provide your feedback? Thanks.

patil-suraj commented on src/transformers/models/bartpho/tokenization_bartpho.py:
```diff
@@ -15,48 +15,40 @@
-    Adapted from [`XLMRobertaTokenizer`]. Based on [SentencePiece](https://github.com/google/sentencepiece).
 
-    This tokenizer inherits from [`PreTrainedTokenizer`] which contains most of the main methods. Users should refer to
+class BartphoTokenizer(XLMRobertaTokenizer):
```
We avoid doing such inheritance in tokenizers, to keep the tokenizers independent from one another. So I would suggest to not inherit from `XLMRobertaTokenizer`. cc @LysandreJik @sgugger

sgugger replied:
Agreed, as mentioned in my comment above :-)

patil-suraj commented on src/transformers/models/bartpho/tokenization_bartpho_fast.py:
```diff
@@ -0,0 +1,138 @@
+# fmt: on
+
+
+class BartphoTokenizerFast(XLMRobertaTokenizerFast):
```
Same comment as above, I would keep it independent from `XLMRobertaTokenizerFast`

sgugger reviewed (commented):
Thanks a lot for your PR! As mentioned by @patil-suraj already, we don't rely on inheritance in Transformers but each object should be fully defined in their configuration/modeling/tokenizet file (there are some instances of subclasses for older models, but this will be cleaned up in the future).

So you should revert your changes in the slow tokenizer file to not inherit on XLMRobertaTokenizer, and make the fast version be independent of XLM-RoBERTa as well.

datquocnguyen committed:
Revise BARTpho slow and fast tokenizers to be independent

datquocnguyen committed:
Fix formatting

datquocnguyen committed:
Fix formatting

datquocnguyen committed:
Fix formatting

datquocnguyen committed:
Merge branch 'main' into main

datquocnguyen committed:
Fix formatting

datquocnguyen commented:
Hi @patil-suraj and @sgugger I revised the slow and fast BartphoTokenizer variants to satisfy your requirements.
Please have a look and give feedback. Thanks. 
cc: @SaulLu @LysandreJik

sgugger reviewed (commented):
Thanks, but it looks like the changes in the slow tokenizer are breaking, which we can't really do.

datquocnguyen committed:
Update src/transformers/models/bartpho/tokenization_bartpho_fast.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

datquocnguyen commented:
> Thanks, but it looks like the changes in the slow tokenizer are breaking, which we can't really do.

@sgugger what do you mean by "we can't really do"? 

The strategy I constructed for my original slow BARTpho tokenizer makes it impossible to have a fast version (https://github.com/huggingface/transformers/pull/13788). So, I adjusted the pre-trained XLM-R sentencepiece model to match with our monolingual Vietnamese vocabulary, and rewrote the slow BARTpho tokenizer, testing to make sure that the  outputs of the original and new ones are the same. In this manner, I could thus have the fast variant.

datquocnguyen committed:
Fix formatting

SaulLu commented on src/transformers/convert_slow_tokenizer.py:
```diff
@@ -551,6 +551,60 @@ def post_processor(self):
         )
 
 
+class BartphoConverter(SpmConverter):
```
I would find it really great if we could avoid having hardcoded values here. Personally, I'll probably see it more clearly once you've answered the two previous questions.

One change you could make to avoid hardcoded values is to replace :
```python
    def unk_id(self, proto):
        unk_id = 3
        return unk_id
```

with 
```python
    def unk_id(self, proto):
        unk_id = self.original_tokenizer.unk_token_id
        return unk_id
```

SaulLu reviewed (commented):
Thank you very much for your contribution! 

I think I personally lack context on what motivated the changes in the python version of the `BartphoTokenizer` tokenizer. In particular, I understand that you changed the spm model uploaded to the hub to `vinai/bartpho-syllable` (before it had a `250000`-sized vocabulary and now it has a `40003`-sized vocabulary).

Additionally, those changes are breaking for the slow tokenizer and we generally try to avoid those in `transformers` (cc @LysandreJik , @sgugger and @patil-suraj ) :smile:

datquocnguyen commented on src/transformers/models/bartpho/tokenization_bartpho.py:
```diff
@@ -151,27 +147,30 @@ def __init__(
             **kwargs,
         )
 
-        self.vocab_file = vocab_file
-        self.monolingual_vocab_file = monolingual_vocab_file
```
The "new" spm model, producing the same tokenization output as the "old" 250K-sized model, now can represent the monolingual Vietnamese vocabulary without the `monolingual_vocab_file` file.

For example, the following two functions return the same results.

```
def _convert_id_to_token(self, index):
         """Converts an index (integer) in a token (str) using the vocab."""
         if index in self.fairseq_ids_to_tokens:
             return self.fairseq_ids_to_tokens[index]
         return self.sp_model.IdToPiece(index - self.fairseq_offset)
         
def _convert_id_to_token(self, index):
         """Converts an index (integer) in a token (str) using the vocab."""
         return self.fairseq_ids_to_tokens[index]
 ```

SaulLu replied:
For my part, I would just like to point out why I think it is complicated to have a tokenizer fast equivalent to slow (proposed in your PR #13788) and why I think the new version is not equivalent to the one proposed before.

What's special about your slow tokenizer is that you tokenize plus you set ids. If the id does not belong to your monolingual vocab then you replace it with the id of the unknown token. This logic is implemented here: 
https://github.com/huggingface/transformers/blob/ade7371a41145c53fe90f66da7e5606f96068e98/src/transformers/models/bartpho/tokenization_bartpho.py#L284-L289

This logic has a particular effect when the input text contains tokens present in the `vocab_file` but not in the `monolingual_vocab_file`.

Indeed, with the version of transformers on master:
```python
from transformers import BartphoTokenizer

model_name = "vinai/bartpho-syllable"
tokenizer_s = BartphoTokenizer.from_pretrained(model_name, revision="e237566c590483a1bdd03aba4687d7516d091e0f")

text = " 접수居家"
tokenizer_s.convert_ids_to_tokens(tokenizer_s.encode(text))
```
gives
```python
['<s>', '<unk>', '<unk>', '</s>']
```

Note: I apologize if VV means something weird, I just concatenated 2 tokens to illustrate my previous point. I have no idea what it means (or doesn't mean).

With the new fast version you propose, here is what it would look like:
```python
from transformers import BartphoTokenizerFast

model_name = "vinai/bartpho-syllable"
tokenizer_f = BartphoTokenizerFast.from_pretrained(model_name)

text = " 접수居家"
tokenizer_f.convert_ids_to_tokens(tokenizer_f.encode(text))
```
gives
```python
['<s>', '▁', '<unk>', '수', '居', '家', '</s>']
```

I hope it makes sense! :smile:

datquocnguyen replied:
@SaulLu I realized this. To develop the fast tokenizer (tokenizer_file), I had to hack into the pre-trained XLM-R sentencepiece model to reduce its vocab size from 250k to 40k (this is for aligning with the existing monolingual vocabulary). Thus, it might affect the BPE subword segmentation/tokenization.

In particular, on [a monolingual Vietnamese corpus](https://drive.google.com/file/d/1wyOW6dXM2iphVhj-8K0LU_XAQPuETkM5/view?usp=sharing) of 590K sentences (~18M tokens, i.e. the averaged sentence length is **30** tokens), the fast and the original slow tokenizers produce different subword tokenization outputs for only 14.5K sentences (rate: 14.5K / 590K ~ **1 / 40**). Note that the difference in subword tokenization mismatch, in most cases, is accounted for only one token (genearlly, less appearing token) in each of the 14K sentences. So the token-level rate for appearing a subword tokenization mismatch between the fast and slow tokenizers is about 1 / (40 * 30) = 1 / 1200, very minor, not really affecting model's performance, I believe.

datquocnguyen replied:
```python
count1, count2 = 0, 0
with open("Vietnamese_file_path", "r") as infile:
    for line in infile:
        count1 += 1

        tokenIDs_f = tokenizer_f(line.strip())['input_ids']
        tokenIDs_s = tokenizer_s(line.strip())['input_ids']
        
        if tokenIDs_f != tokenIDs_s:
            count2 += 1
```

datquocnguyen replied:
Btw, the text in your example is Chinese, so it is not surprising that you observed the subword tokenization mismatch. @SaulLu

datquocnguyen replied:
The same very minor mismatch rate for BPE subword tokenization is also observed on [another Vietnamese dataset](https://drive.google.com/file/d/1mhp0BMNkZq9SBxeNYrJx08BeNRerzpnl/view?usp=sharing) of about 9M sentences. @SaulLu 

> @SaulLu I realized this. To develop the fast tokenizer (tokenizer_file), I had to hack into the pre-trained XLM-R sentencepiece model to reduce its vocab size from 250k to 40k (this is for aligning with the existing monolingual vocabulary). Thus, it might affect the BPE subword segmentation/tokenization.
> 
> In particular, on [a monolingual Vietnamese corpus](https://drive.google.com/file/d/1wyOW6dXM2iphVhj-8K0LU_XAQPuETkM5/view?usp=sharing) of 590K sentences (~18M tokens, i.e. the averaged sentence length is **30** tokens), the fast and the original slow tokenizers produce different subword tokenization outputs for only 14.5K sentences (sentence-level mismatch rate: 14.5K / 590K ~ **1 / 40**). Note that the difference in subword tokenization mismatch, in most cases, is accounted for only one token (genearlly, less appearing token) in each of the 14K sentences. So the token-level rate for appearing a subword tokenization mismatch between the fast and slow tokenizers is about 1 / (40 * 30) = 1 / 1200, very minor, not affecting model's performance, I believe.

datquocnguyen replied:
I think the tiny different rate between tokenization outputs of the original slow and new fast tokenizers confirms that the new fast tokenizer is equivalent to the original slow one. 
I am wondering what your thought is w.r.t. this? Thank you @SaulLu

datquocnguyen committed:
Remove hardcoded value

SaulLu commented on src/transformers/models/bartpho/tokenization_bartpho.py:
```diff
@@ -151,27 +147,30 @@ def __init__(
 
-        # Load the reduced vocab
-
-        # Keep order of special tokens for backward compatibility
-        self.fairseq_tokens_to_ids = {}
-        cnt = 0
-        for token in [bos_token, pad_token, eos_token, unk_token, sep_token, cls_token]:
-            if str(token) not in self.fairseq_tokens_to_ids:
-                self.fairseq_tokens_to_ids[str(token)] = cnt
-                cnt += 1
-        with open(monolingual_vocab_file, "r", encoding="utf-8") as f:
-            for line in f.readlines():
-                token = line.strip().split()[0]
-                self.fairseq_tokens_to_ids[token] = len(self.fairseq_tokens_to_ids)
-        if str(mask_token) not in self.fairseq_tokens_to_ids:
-            self.fairseq_tokens_to_ids[str(mask_token)] = len(self.fairseq_tokens_to_ids)
```
Thanks for the explanations! 

To make sure I understand everything, why it was not satisfactory to use `MBartTokenizer`?

```python
model_name = "vinai/bartpho-syllable"
tokenizer_mbart = MBartTokenizer.from_pretrained("vinai/bartpho-syllable")
```

datquocnguyen replied:
A `fairseq` BARTpho sequence has the following format for a single sentence X: `<s> X </s>`, while MBartTokenizer will produce `X </s> src_lang_code`.

datquocnguyen replied:
I am wondering whether you would have any further concern regarding this PR ? @SaulLu 
(cc @LysandreJik , @sgugger , @patil-suraj and @patrickvonplaten )

SaulLu replied:
Thank you for the previous explanations.

For my part, I think my main feedback is [this one](https://github.com/huggingface/transformers/pull/17254#discussion_r876742460)

From my point of view, it may be possible to make a fast tokenizer but it will probably take some hacking to get the same tokenization strategy.

datquocnguyen replied:
@SaulLu Thank you very much for feedback. 
My response to your feedback is at [here](https://github.com/huggingface/transformers/pull/17254#discussion_r876864300)

datquocnguyen replied:
@SaulLu I revised my code so that now the fast tokenizer has the same tokenization strategy as of the slow original one, doing some hacking in the function `_convert_encoding`. 

In particular, I extend the monolingual vocab with all other tokens from the multilingual vocab. 

```python
class BartphoConverter(SpmConverter):
    def vocab(self, proto):

        dict = {piece.piece: piece.score for piece in proto.pieces[3:]}

        # Original monolingual vocab where `mask_token_id = len(vocab) - 1` 
        vocab = [
            (k, dict[k] if k in dict.keys() else 0.0) for k, v in self.original_tokenizer.fairseq_tokens_to_ids.items()
        ]

       # Extend the monolingual vocab with all remaining tokens from the pretrained XLM-R sentencepiece model.
        for piece in proto.pieces[3:]:
            if piece.piece not in self.original_tokenizer.fairseq_tokens_to_ids.keys():
                vocab += [(piece.piece, piece.score)]

        return vocab
```

The trick is to map the token with its id that is larger than `mask_token_id` into `unk_token_id` in the function `_convert_encoding` of the fast tokenizer: https://github.com/datquocnguyen/transformers/blob/1b22570797d8e70de0b6dabccf8c9c4ececff3b9/src/transformers/models/bartpho/tokenization_bartpho_fast.py#L175 

```python
        for e in encodings:
            #encoding_dict["input_ids"].append(e.ids)
            ids = [id if id <= self.mask_token_id else self.unk_token_id for id in e.ids]
            encoding_dict["input_ids"].append(ids)
```
cc: @sgugger FYI, I keep the original slow tokenizer intact. Changes now do not break the existing code.

datquocnguyen replied:
@SaulLu I revised my code so that now the fast tokenizer has the same tokenization strategy as of the slow original one, doing some hacking in the function `_convert_encoding`. 

In particular, I extend the Vietnamese monolingual vocab with all other tokens from the multilingual vocab. So now I have a variant of the multilingual vocab, where the first half stores all tokens from the Vietnamese monolingual vocab and the second half stores the rest, separated by the "mask_token". 

```python
class BartphoConverter(SpmConverter):
    def vocab(self, proto):

        dict = {piece.piece: piece.score for piece in proto.pieces[3:]}

        # Original monolingual vocab where `mask_token_id = len(vocab) - 1` 
        vocab = [
            (k, dict[k] if k in dict.keys() else 0.0) for k, v in self.original_tokenizer.fairseq_tokens_to_ids.items()
        ]

       # Extend the monolingual vocab with all remaining tokens from the pretrained XLM-R sentencepiece model.
        for piece in proto.pieces[3:]:
            if piece.piece not in self.original_tokenizer.fairseq_tokens_to_ids.keys():
                vocab += [(piece.piece, piece.score)]

        return vocab
```

The trick is to map the token with its id that is larger than `mask_token_id` into `unk_token_id` in the function `_convert_encoding` of the fast tokenizer: https://github.com/datquocnguyen/transformers/blob/1b22570797d8e70de0b6dabccf8c9c4ececff3b9/src/transformers/models/bartpho/tokenization_bartpho_fast.py#L175 

```python
        for e in encodings:
            #encoding_dict["input_ids"].append(e.ids)
            ids = [id if id <= self.mask_token_id else self.unk_token_id for id in e.ids]
            encoding_dict["input_ids"].append(ids)
```
cc: @sgugger FYI, I keep the original slow tokenizer intact. Changes now do not break the existing code.

datquocnguyen commented:
Please note that the unsuccessful checks are due to the failed `test_modeling_wav2vec2_conformer.py`, not related to our BartphoTokenizer. @SaulLu

patrickvonplaten commented:
> Please note that the unsuccessful checks are due to the failed `test_modeling_wav2vec2_conformer.py`, not related to our BartphoTokenizer. @SaulLu

@SaulLu fixed the wav2vec2_conformer tests on master

sgugger commented:
@datquocnguyen We can't merge anything that has any breaking change on the existing tokenizer, as I said before.

datquocnguyen commented:
@sgugger Ah, I now see your point. I initially thought the code would be much nicer if I also push a new version of the slow tokenizer. But then it breaks the existing code. 

Anyway, the fast tokenizer would totally work without changing the original code of the slow tokenizer (as I already developed the fast_tokenizer_file), I think. I would need a bit of time to roll back the slow tokenizer to its original version. 
 
(cc @SaulLu , @LysandreJik , @patil-suraj and @patrickvonplaten )

datquocnguyen committed:
Revert the new slow tokenizer to the original slow one

datquocnguyen committed:
Fix formatting

datquocnguyen committed:
The fast tokenizer with the same tokenization strategy as the slow one

datquocnguyen committed:
Fix formatting

datquocnguyen commented:
Please note that the failed tests are because of: 

1. tests/models/wav2vec2_conformer/test_modeling_wav2vec2_conformer.py::Wav2Vec2ConformerModelTest::test_save_load_fast_init_from_base, not related to BartphoTokenizer
2. I am not quite sure why `tokenization_bartpho_fast.py` could pass your `black` code-quality test pipeline, while it succeded in my local machine. 

<img width="941" alt="Screen Shot 2022-05-21 at 6 20 11 pm" src="https://user-images.githubusercontent.com/2412555/169649501-ff26a303-0325-430c-9391-603f56f85e7d.png">

datquocnguyen committed:
Add fast tokenizers for PhoBERT and BERTweet

datquocnguyen committed:
Fix formatting

datquocnguyen committed:
Add require_torch

datquocnguyen commented:
Hi @SaulLu , @sgugger , @patil-suraj @LysandreJik and @patrickvonplaten

In addition to a fast BARTpho tokenizer, I also revised my code to add fast tokenizers for BERTweet and PhoBERT. Here, changes now do not break existing slow tokenizers. My hacking trick to have the same tokenization strategy for both slow and fast variants is already mentioned [here](https://github.com/huggingface/transformers/pull/17254#discussion_r878687089).

Please have a look and provide feedback. Thanks!

Note that I have no idea to fix the failed test `check_code_quality` w.r.t. `black`:

```
#!/bin/bash -eo pipefail
black --check --preview examples tests src utils
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install black[jupyter]``
would reformat src/transformers/models/bartpho/tokenization_bartpho_fast.py

Oh no! 💥 💔 💥
1 file would be reformatted, 1594 files would be left unchanged.

Exited with code exit status 1
```

However, the target file "tokenization_bartpho_fast.py" is left unchanged in my local machine:

<img width="1016" alt="Screen Shot 2022-05-22 at 11 21 42 pm" src="https://user-images.githubusercontent.com/2412555/169705603-c5dbb397-9a63-4ce8-904d-fff1280c07bf.png">

I think there might be an inconsistency with `black` used in my local machine and in your CI, so I could not fix it from my side. It would be great if you guys could help fix it. Thanks a lot.

SaulLu reviewed (commented):
Thanks for reworking this PR to propose a solution compatible with the slow tokenizer. I think we are getting closer to a compatible solution. :smile: 

For the moment I have only looked at the changes concerning `Bartpho` (not being familiar with the other tokenizers for which you propose fast versions).

I've left a few comments but overall I think my biggest point is that with a hack like this - as the `test_add_tokens_tokenizer` test shows - we're really going to have to make sure that all the tests in `test_tokenization_common.py` test what we think they should. Indeed, not all of these tests will be robust to the proposed hack, unfortunately. For example, for the moment the proposed hack does not allow to use `add_tokens` feature as intended. To solve this problem, perhaps we could try to choose another rule to assign the unknow id?

I think it would also help to have an additional test with hard-coded target values that tests all the public methods in the tokenizer on examples that would behave specifically with your tokenizer (i.e. a text that contains tokens that would be in `monolingual_vocab_file` and others that would only be in `vocab_file` and perhaps a token that is not in any vocabulary).

datquocnguyen commented:
@SaulLu Thank you very much for your detailed feedback and suggestion. Before moving forward to revise the code w.r.t. the `add_tokens` feature, it would be great if you could provide some more context/clarification on the intention of using `add_tokens`.

For (monolingual) Vietnamese where the Vietnamese lexicon of syllables contains 8K syllable types, using a monolingual vocab of 40K types in `vinai/bartpho-syllable` is far more than enough to cover all possible cases of Vietnamese syllables. I am currently not sure whether the `add_tokens` feature is needed in our tokenizer/model when using our tokenizer/model on Vietnamese data?

SaulLu commented:
> @SaulLu Thank you very much for your detailed feedback and suggestion. Before moving forward to revise the code w.r.t. the add_tokens feature, it would be great if you could provide some more context/clarification on the intention of using add_tokens.

@datquocnguyen I think there are many, many use cases for `add_tokens`. But for example, we can imagine a use case where a user would like to fine-tune the model on a task that needs to identify specific tokens: like for example `"<QUESTION>"` and `"<ANSWER>"`. This method is convenient because it is unified across all tokenizers.

datquocnguyen committed:
Improved tokenization strategy for BartphoTokenizerFast

datquocnguyen committed:
Original BERTweet and PhoBERT tokenizers

datquocnguyen committed:
Fix format

datquocnguyen commented on src/transformers/convert_slow_tokenizer.py:
```diff
@@ -551,6 +651,36 @@ def post_processor(self):
+
+    def unk_id(self, proto):
+        unk_id = 3
+        return unk_id
```
I updated the code to avoid hardcoded values

datquocnguyen commented on src/transformers/models/bartpho/tokenization_bartpho_fast.py:
```diff
@@ -0,0 +1,276 @@
+        if os.path.abspath(self.tokenizer_file) != os.path.abspath(out_tokenizer_file):
+            copyfile(self.tokenizer_file, out_tokenizer_file)
+
+        return (out_vocab_file, out_tokenizer_file)
```
Yes, I fixed this issue.

datquocnguyen commented:
@SaulLu Thank you very much for your feedback. 

I improved the hacking strategy to handle the issue with newly added tokens. 

Assume that the sizes of the multilingual and monolingual vocabularies are X and Y, respectively (here, X > Y, and Y is set at `mask_token_id` in our hacking strategy). Added tokens A1, A2, A3, ... would have original ids of X, X+1, X+2,... that will be mapped into new ids Y, Y+1, Y+2,...

I extended the original function `get_added_vocab` into `get_added_vocab_hacking` to extract a dictionary `added_vocab ` {A1: Y, A2: Y+1, A3: Y+2, ...} and another dictionary `id_mapping` of id mapping  {X: Y, X+1: Y+1, X+2: Y+2, ...}

```python
    def get_added_vocab_hacking(self):
        """
        Returns the added tokens in the vocabulary as a dictionary of token to index.

        Returns:
            `Dict[str, int], Dict[int, int]`: The added tokens, and their original and new ids
        """
        base_vocab = self._tokenizer.get_vocab(with_added_tokens=False)
        full_vocab = self._tokenizer.get_vocab(with_added_tokens=True)
        base_vocab_size = len(base_vocab)
        if len(full_vocab) == base_vocab_size:
            return {}, {}


        # Tokens in added_vocab should have ids that are equal to or larger than the size of base_vocab
        added_vocab = dict(
            (tok, index + 1 - base_vocab_size + self.mask_token_id)
            for tok, index in full_vocab.items()
            if tok not in base_vocab
        )


        id_mapping = dict((index, full_vocab[tok]) for tok, index in added_vocab.items())


        return added_vocab, id_mapping

```

So in tokenization, the previous strategy maps all ids larger than `mask_token_id` to `unk_token_id` now is revised to also handle added tokens [as follows](https://github.com/datquocnguyen/transformers/blob/f59b4afeb1af6551feac5d3214bbdf582ebbb098/src/transformers/models/bartpho/tokenization_bartpho_fast.py#L234-L242):


```python

            ids = []
            for (id, token) in zip(e.ids, e.tokens):
                if id <= self.mask_token_id:
                    ids.append(id)
                else:
                    if token.strip() in added_vocab: #  handle added tokens
                        ids.append(added_vocab[token.strip()])
                    else:
                        ids.append(self.unk_token_id)
```

In addition, [a preprocess of mapping ids](https://github.com/datquocnguyen/transformers/blob/f59b4afeb1af6551feac5d3214bbdf582ebbb098/src/transformers/models/bartpho/tokenization_bartpho_fast.py#L174-L197) Y, Y+1, Y+2, ... into X, X+1, X+2 is applied before decoding:

```python
    def _decode(
        self,
        token_ids: Union[int, List[int]],
        skip_special_tokens: bool = False,
        clean_up_tokenization_spaces: bool = True,
        **kwargs
    ) -> str:
        self._decode_use_source_tokenizer = kwargs.pop("use_source_tokenizer", False)


        if isinstance(token_ids, int):
            token_ids = [token_ids]


        # Mapping added tokens' ids into their original values
        _, id_mapping = self.get_added_vocab_hacking()
        if len(id_mapping) > 0:
            token_ids = [id_mapping[id] if id in id_mapping else id for id in token_ids]


        text = self._tokenizer.decode(token_ids, skip_special_tokens=skip_special_tokens)


        if clean_up_tokenization_spaces:
            clean_text = self.clean_up_tokenization(text)
            return clean_text
        else:
            return text
```

With this improved strategy, there are two tests needed to override:

```python
    def test_tokenizer_fast_store_full_signature(self):
        """
        Override the original test as BartphoTokenizer requires a monolingual_vocab_file rather than a merges_file
        """
```

```python
def test_add_tokens_tokenizer(self):
        """
        Override the original test as in the fast tokenizer, the actual vocab_size is in fact mask_token_id + 1
        """
```

datquocnguyen commented on tests/models/bartpho/test_tokenization_bartpho.py:
```diff
@@ -43,25 +49,155 @@ def setUp(self):
+                self.assertEqual(added_toks, len(new_toks))
+                self.assertEqual(all_size_2, all_size + len(new_toks))
+
+                tokens = tokenizer.encode("aaaaa bbbbbb low cccccccccdddddddd l", add_special_tokens=False)
```
@SaulLu the new strategy [HERE](https://github.com/huggingface/transformers/pull/17254#issuecomment-1139492485) now handles this successfully.

datquocnguyen commented on tests/models/bartpho/test_tokenization_bartpho.py:
```diff
@@ -43,25 +49,155 @@ def setUp(self):
+        input_bpe_tokens = [4, 5, 6, 7, 8, 3]
         self.assertListEqual(tokenizer.convert_tokens_to_ids(input_tokens), input_bpe_tokens)
+
+    def test_rust_and_python_full_tokenizers(self):
```
@SaulLu the new strategy (https://github.com/huggingface/transformers/pull/17254#issuecomment-1139492485) now helps avoid overriding 3/4 tests (except `test_add_tokens_tokenizer` where in the fast tokenizer, the actual vocabulary size is in fact `mask_token_id + 1`).

datquocnguyen commented on tests/models/bartpho/test_tokenization_bartpho.py:
```diff
@@ -43,25 +49,155 @@ def setUp(self):
+        out-of-monolingual-vocabulary tokens, prevents it from training a new one from scratch.
+        """
+
+        pass
```
@SaulLu I am not quite sure I understand how to do it as I was asked to not break existing code (here, `test_tokenization_common`). 
We should simply use "pass" in those tests, should not we ?  as other tokenizers also use it, e.g. https://github.com/datquocnguyen/transformers/blob/f59b4afeb1af6551feac5d3214bbdf582ebbb098/tests/models/openai/test_tokenization_openai.py#L133-L135

    # tokenizer has no padding token
    def test_padding_different_model_input_name(self):
        pass

datquocnguyen commented on tests/models/bartpho/test_tokenization_bartpho.py:
```diff
@@ -43,25 +49,155 @@ def setUp(self):
+        BartphoTokenizer involves an external monolingual vocabulary file. Thus, it would not pass the original test.
+        """
+
+        pass
```
Those tests are now passed without overriding.

datquocnguyen commented:
@SaulLu please help to review [the imporved strategy](https://github.com/huggingface/transformers/pull/17254#issuecomment-1139492485) and give feedback. Thanks a lot. 

Please note that failed checks  are not related to my bartpho tokenizer, except for one check, however it was successful in my local computer, as detailed at https://github.com/huggingface/transformers/pull/17254#issuecomment-1133932067 
```
#!/bin/bash -eo pipefail
black --check --preview examples tests src utils
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install black[jupyter]``
would reformat src/transformers/models/bartpho/tokenization_bartpho_fast.py

Oh no! 💥 💔 💥
1 file would be reformatted, 1594 files would be left unchanged.

Exited with code exit status 1

```

datquocnguyen commented:
> Hi @datquocnguyen. It's amazing that you added those two new fast tokenizers. However we need PRs to be focused on one thing. Would you terribly mind splitting it in three (one for BARTpho, one for PhoBERT and one for BERTweet)?
> 
> Thanks a lot!

@sgugger I changed the code, so that this PR is only for BARTpho. cc: @SaulLu

datquocnguyen committed:
Improve get_added_vocabulary_hacking

datquocnguyen committed:
Merge pull request #1 from datquocnguyen/main

Improve get_added_vocabulary_hacking

datquocnguyen committed:
Fix formatting

datquocnguyen committed:
Fast tokenizers for PhoBERT and BERTweet

datquocnguyen commented:
@SaulLu please help to review [the improved strategy](https://github.com/huggingface/transformers/pull/17254#issuecomment-1139492485) and give feedback. Thank you very much.

Please note that failed checks  are not related to my bartpho tokenizer, except for one check using `black`, however `black` was successful in my local computer, as detailed at https://github.com/huggingface/transformers/pull/17254#issuecomment-1133932067. Could you provide information about `black` used in your CI?, so I can replicate the issue on my local computer, then fix it. Thanks. 

cc: @sgugger 

```
#!/bin/bash -eo pipefail
black --check --preview examples tests src utils
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install black[jupyter]``
would reformat src/transformers/models/bartpho/tokenization_bartpho_fast.py

Oh no! 💥 💔 💥
1 file would be reformatted, 1594 files would be left unchanged.

Exited with code exit status 1

```

sgugger commented:
You need to install `black==22.3` to have the same results as the CI.

datquocnguyen committed:
Fix formatting

datquocnguyen committed:
Fix formatting

datquocnguyen commented:
@sgugger Following https://github.com/huggingface/transformers/pull/17254#issuecomment-1133932067, I was not aware that I have to include `--preview` into my command `black -l 119 <py_file_path>`. The code quality check is passed. 

There are now 4 failed checks not caused by BartphoTokenizerFast, I believe:

- FAILED tests/models/layoutlmv2/test_tokenization_layoutlmv2.py::LayoutLMv2TokenizationTest::test_saving_tokenizer_trainer
====== 1 failed, 135 passed, 32 skipped, 20 warnings in 142.09s (0:02:22) ======
- FAILED tests/pipelines/test_pipelines_summarization.py::SummarizationPipelineTests::test_small_model_pt
- `run_tests_flax` = 804 failed, 5364 passed, 11260 skipped, 7960 warnings in 1306.16s (0:21:46) ==
- `run_tests_torch` = 823 failed, 10738 passed, 6752 skipped, 5425 warnings in 1554.96s (0:25:54) ==

cc: @SaulLu , @LysandreJik, @patil-suraj and @patrickvonplaten It would be great if you guys can also help review this PR. Thanks a lot.

sgugger commented:
You will need to rebase on the main branch to fix the test failures. It's due to the botched release of Protobuf that breaks everything (the main branch has it pinned).

datquocnguyen committed:
Merge pull request #2 from huggingface/main

Rebase on the main branch

datquocnguyen committed:
Merge pull request #4 from huggingface/main

Rebase on the main branch

datquocnguyen commented:
@sgugger I rebased the main branch with the latest commits from `transformers`. 

There are 3 failed checks not relevant to the BartphoTokenizer:

- `Build PR Documentation / build / build_pr_documentation`: urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
- `run_tests_tf`: FAILED tests/models/mobilebert/test_modeling_tf_mobilebert.py::TFMobileBertModelTest::test_resize_token_embeddings
- `run_tests_torch`: FAILED tests/models/glpn/test_feature_extraction_glpn.py::GLPNFeatureExtractionTest::test_call_pytorch

fyi, @SaulLu , @LysandreJik, @patil-suraj and @patrickvonplaten

LysandreJik commented:
Hey @datquocnguyen, thanks a lot for your PR and for working hard on this! I think this is one situation where the code on the hub (detailed below) would fit really well, for the following reasons:

- The tokenizer code that is defined seems to work with the vocabulary you have worked with so far, but is unlikely to work with other vocabularies. At least it won't be the case unless the approach you have taken to generate that vocabulary is very documented, as it is very complex when compared to other tokenizers.
- If I understand correctly, the approach taken could have been handled by a single vocabulary rather than 2. I definitely understand why doing it like this for BARTpho makes sense, but this is unlikely to be the case for other checkpoints leveraging this architecture.
- The code in `transformers` has to be maintained for years, so we want to optimize for heavily tested code; the methods that you add, while definitely useful in order to get the BARTpho tokenization right, are not tested and have `hacking` in their name, which shows that they're targetting something a bit different than what we aim to solve with `transformers`' internal code (but that definitely has its place on the hub!).

Finally, you're moving very fast with your implementations, which is great. However, given the backwards-compatibility approach we have chosen and the fact that we want production-ready code means that we'll be slowing things down in this case, unfortunately.

---

The code on the hub is explained [here](https://huggingface.co/docs/transformers/custom_models). It's a way to share models and configurations by sharing their modeling code directly on the hub. When doing `from_pretrained`, you can then fetch the code on the hub. BARTpho is exactly the kind of use-cases we had in mind when working on this feature - we just didn't get to implementing the tokenizer code yet! I think we should work to enable this ASAP and have BARTpho be a first trial.

This would enable you to move as fast as you need, while providing the same functionality to downstream `transformers` users, and will allow you to manage your repositories as you see fit. Would that work for you?

datquocnguyen commented:
[@LysandreJik](https://github.com/LysandreJik) Thanks for your detailed feedback.
Before I go to answer whether the code on the hub would work for me.
I am just concerning your first comment:

> The tokenizer code that is defined seems to work with the vocabulary you have worked with so far, but is unlikely to work with other vocabularies. At least it won't be the case unless the approach you have taken to generate that vocabulary is very documented, as it is very complex when compared to other tokenizers.

So I would try to respond to this first comment. As detailed in [#13788 (comment)](https://github.com/huggingface/transformers/pull/13788#issuecomment-931908671), regarding the use case of BartphoTokenizer: Other languages can thus simply reuse BartphoTokenizer with their `monolingual_vocab_file`. The goal is to reduce the model sizes of existing pre-trained XLM-RoBERTa/mBART models when applying to a smaller set of languages instead of the whole 50/100 languages. Here, you would trim XLM-RoBERTa/mBART to just dealing with subwords in the `monolingual_vocab_file` while not requiring retraining the sentencepiece tokenization model.

The generation process of BARTpho vocabulary is not that very complicated, as detailed in [#13788 (comment)](https://github.com/huggingface/transformers/pull/13788#issuecomment-931908671). In particular, I apply a pre-trained sentencepiece tokenization model from a pre-trained language model (e.g., XLM-RoBERTa/mBART/...) to segment sentences in a language/task-specific corpus, and then selected just top X (e.g. 40K) subwords to be included in a specific vocabulary for my downstream language/task (here, I named this specific vocabulary as `monolingual_vocab_file`).

Regarding BartphoTokenizerFast, the process of generating the `tokenizer_file` is that: (1) I load the slow BartphoTokenizer, (2) call the function `convert_slow_tokenizer` to convert it into a fast variant, and (3) then save the fast one. This might be a bit complicated for others as it is not well-documented, but I could simply abandon the use of `tokenizer_file` in BartphoTokenizerFast. Thus BartphoTokenizerFast would just create and convert a slow tokenizer BartphoTokenizer to build the backend.

I believe there are many use cases in which BartphoTokenizer/BartphoTokenizerFast would fit.

[@SaulLu](https://github.com/SaulLu) As you have been playing around with BartphoTokenizer, is there any comment from your side regarding [@LysandreJik](https://github.com/LysandreJik)' first point. Thank you both.

datquocnguyen commented:
@LysandreJik I did dig into the code on the hub, and am wondering whether I understand your approach correctly:

- Instead of merging `tokenization_bartpho_fast.py` into the main `transformers` branch, we now just need to upload/push it to `https://huggingface.co/vinai/bartpho-syllable/tree/main`. 

- There would be an upcoming feature of `sharing a custom tokenizer`, which I should register BartphoTokenizerFast from `vinai/bartpho-syllable` or `https://huggingface.co/vinai/bartpho-syllable/blob/main/tokenization_bartpho_fast.py`. Then it would allow users to automatically download or import `tokenization_bartpho_fast.py` and use BartphoTokenizerFast via AutoTokenizer with existing features in the main `transformers` branch. 

So what I should do is to wait until you guys complete that `sharing a custom tokenizer` feature and then I would just need to have some piece of code for registering BartphoTokenizerFast with `register_for_auto_class('AutoTokenizer')` and it would run as the same as merged into the main `transformers` branch, wouldn't it? 

Thanks.

cc: @SaulLu

datquocnguyen committed:
Merge pull request #5 from huggingface/main

Update the latest commits from the main transformers branch

datquocnguyen committed:
Merge pull request #6 from huggingface/main

Update the latest commits from the main transformers branch

datquocnguyen committed:
Merge pull request #7 from huggingface/main

Update with the latest commits from transformers

datquocnguyen commented:
For a wider context where some subwords appearing in the "merges" file do not appear in the "vocab" file as in CTRL, FlauBERT, PhoBERT and BERTweet and the like (i.e. we would need to convert those subwords into unkn_id during encoding), it is likely impossible to develop a fast tokenizer variant using documented approaches while keeping the same tokenization strategy.

Thus, the trick used in BartphoTokenizerFast would come into play, and help solve this issue. If merged, it is then straightforward to develop similar fast tokenizers for CTRL, FlauBERT, PhoBERT and BERTweet.

It would be great if @LysandreJik @SaulLu @patrickvonplaten or @sgugger could provide concrete feedback on whether this PR will have a chance to be merged. If this PR could not be merged, then what is the status of the "sharing a custom tokenizer on the hub" feature (e.g. tentative date for releasing this feature) ?

Thank you very much.

datquocnguyen committed:
Merge pull request #8 from huggingface/main

Update the latest commits

SaulLu commented:
Hi @datquocnguyen ,
I echo [Lysandre's answer](https://github.com/huggingface/transformers/pull/17254#issuecomment-1143221043): I thank you for working very hard for this PR :hugs: and I also think it would be a very good fit for the feature on the hub. And this addition will be really useful for the community!

> It would run as the same as merged into the main transformers branch, wouldn't it?

Yes, the idea is that it would be (almost) identical to what you have with transformers! I don't know when it will be released (as I'm not directly working on it), but it seems to be a high-priority feature!

> For a wider context where many subwords appearing in the "merges" file do not appear in the "vocab" file as in CTRL, FlauBERT, PhoBERT and BERTweet and the like (i.e. slow tokenizers would convert those subwords into unkn_id during encoding), it is likely impossible to develop a fast tokenizer variant using documented approaches while keeping the same tokenization strategy.

Indeed, you raise a very good point. I have also observed that there are tokens listed in the merge rules that do not appear in the vocabulary for `FlauBERT` - and I believe you that this is also the case for `CTRL`, `PhoBERT` and `BERTweet`. Nevertheless, from my point of view, looking at `FlauBERT`'s code, the fix that seems to me the most suitable for our API (tokenizer slow → converter → tokenizer fast ) would be to clean up the merge file during the conversion step. This technique would indeed avoid having to modify the tokenizer fast core method(s). I've attached a snippet below that illustrates this idea. Am I missing something by thinking that this would achieve the desired final behaviour? 

------------
_Snippet to illustrate the merges file "clean up" that I have in mind_

To test this snippet, we need to retrieved locally the `vocab.json` and `merges.txt` files of `FlauBERT`, for example by doing `git clone https://huggingface.co/flaubert/flaubert_base_cased`.

Then, if we try to test to initialize a tokenizer fast (pure without the  transformers's tokenizer wrapper for the moment), we observe that it raises an error
```python
from tokenizers import Tokenizer
from tokenizers.models import BPE
import json

file_vocab = f"flaubert_base_cased/vocab.json"
file_merges = f"content/flaubert_base_cased/merges.txt"

with open(file_vocab) as f:
    vocab = json.load(f)

with open(file_merges) as f:
    merges = f.readlines()

merges = [merge.split(" ") for merge in merges]
merges = [(merge[0], merge[1]) for merge in merges if len(merge)==3]

tokenizer = Tokenizer(
                BPE(
                    vocab,
                    merges,
                    unk_token="<unk>",
                    end_of_word_suffix="</w>",
                    fuse_unk=True,
                )
            )
```
Error message:
```bash
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
[<ipython-input-26-b70d9b50bd17>](https://localhost:8080/#) in <module>()
      5                     unk_token="<unk>",
      6                     end_of_word_suffix="</w>",
----> 7                     fuse_unk=True,
      8                 )
      9             )

Exception: Error while initializing BPE: Token `trouvécap` out of vocabulary
```
But by cleaning the merges file we can initialize the tokenizer without errors
```python
new_merges = []
for token_1, token_2 in merges:
    if token_1 not in vocab or token_2 not in vocab or f"{token_1}{token_2}" not in vocab:
        print(token_1, token_2)
        continue
    new_merges.append((token_1, token_2))
   
tokenizer = Tokenizer(
                BPE(
                    vocab,
                    merges,
                    unk_token="<unk>",
                    end_of_word_suffix="</w>",
                    fuse_unk=True,
                )
            )
```

datquocnguyen committed:
Merge pull request #9 from huggingface/main

Update the latest commits from transformers

datquocnguyen committed:
Merge pull request #10 from huggingface/main

Update the latest commits from v4.20.0

datquocnguyen committed:
Merge pull request #11 from huggingface/main

Merge the latest commits from the original transformers branch

datquocnguyen commented:
I am still looking forward to using the new "sharing a custom tokenizer" feature =)

SaulLu commented:
> Cleaning the "merges" file will definitely result in different encoding outputs from the slow and fast tokenizers. For example, in the case of FlauBERT, the slow and fast tokenizers will encode/tokenize any word containing the sub-string trouvécap differently.

I'm sorry, I didn't react to your message! You are right, my proposal will not be exactly the same as the current slow version.

One specific thing to know about this particular case of FlauBERT is that currently the slow tokenizer doesn't behave exactly like FlauBERT's original tokenizer which used [FastBPE](https://github.com/glample/fastBPE).

For example, `trouvécaptivantes` is not tokenized in the same way:
```
Transformers version:     ['<s>', '<unk>', 'tiv', 'antes</w>', '</s>']
FastBPE version:          ['<s>', 'trouv', 'écap', 'tiv', 'antes</w>', '</s>']
```

Ideally, we would like to have an exact match, but in this case I think the changes that would have to be made to achieve this would be very cumbersome compared to the difference observed (`trouvécaptivantes` is not a word in French but the concatenation of 2 words, without typography we should have had `trouvé captivantes`). All that to say, it's very very complicated to have perfect matching between different tokenization libraries and maintaining long-term hacks is not easy and that's why I think the sharing feature is really a perfect use case for your proposal!

datquocnguyen committed:
Merge pull request #12 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #13 from huggingface/main

Update

datquocnguyen committed:
Merge pull request #14 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #15 from huggingface/main

v4.21.0

datquocnguyen committed:
Merge pull request #17 from huggingface/main

Update

datquocnguyen commented:
> I am still looking forward to using the upcoming "sharing a custom tokenizer" feature =)

https://github.com/datquocnguyen/transformers/tree/fast_tokenizers_BARTpho_PhoBERT_BERTweet also contains fast tokenizer implementations of PhoBERT and BERTweet, which others might find them useful to develop similar fastBPE-based tokenizers for other models such as CTRL & FlauBERT.

datquocnguyen committed:
Merge pull request #18 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #19 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #20 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #21 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #22 from huggingface/main

Update with the latest commits

datquocnguyen committed:
Merge pull request #23 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Fast tokenizers Bartpho Phobert Bertweet

datquocnguyen committed:
Update test_tokenization_bartpho.py

datquocnguyen committed:
Merge pull request #24 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Update test_tokenization_bartpho.py

datquocnguyen committed:
Update test_tokenization_bartpho.py

datquocnguyen committed:
Merge pull request #25 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Update test_tokenization_bartpho.py

datquocnguyen committed:
Merge pull request #27 from datquocnguyen/tmp_branch

Update commits

datquocnguyen committed:
Merge pull request #28 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Update commits

datquocnguyen committed:
Merge pull request #29 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #30 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #31 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Merge pull request #29 from huggingface/main

datquocnguyen committed:
Merge pull request #32 from huggingface/main

Update the latest commits

datquocnguyen committed:
Merge pull request #33 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Update the latest commits

datquocnguyen committed:
Merge pull request #34 from huggingface/main

Update latest commits

datquocnguyen committed:
Merge pull request #35 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Update latest commits

datquocnguyen committed:
Merge pull request #36 from huggingface/main

Update commits

datquocnguyen committed:
Merge pull request #37 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Update commits

datquocnguyen commented:
@SaulLu @LysandreJik Hi, I am just wondering about the status/progress of the "sharing a custom tokenizer" feature on the hub. Will it be released? Thanks.

datquocnguyen committed:
Merge pull request #38 from huggingface/main

Update

datquocnguyen committed:
Merge pull request #39 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Update

datquocnguyen commented:
Hi @SaulLu  @LysandreJik , I am wondering about the status/progress of the "sharing a custom tokenizer" feature on the hub. Is there anything I can help with? This feature would help BERTweet, PhoBERT, BARTpho and the like to be easier to be used with their fast customed tokenizers. Thank you.

LysandreJik commented:
The custom tokenizer should now work correctly! @ArthurZucker, if you have a spare cycle, could you look into supporting the tokenizers added here by @datquocnguyen with code on the hub using the custom tokenizers? 

A guide showing how to is [available here](https://huggingface.co/docs/transformers/custom_models#sending-the-code-to-the-hub). Thanks!

datquocnguyen committed:
Merge pull request #40 from huggingface/main

Update commits

sgugger commented:
It should be one of the `ModelArguments` defined in the example (where the rest of the args, like revision  etc. lie) but yes, I don't see why not!

datquocnguyen commented:
The `ModelArguments` should have `trust_remote_code_model` and `trust_remote_code_tokenizer`  separately for the model and tokenizer loading, respectively, shouldn't it? For example:

```
tokenizer_name_or_path = model_args.tokenizer_name if model_args.tokenizer_name else model_args.model_name_or_path
    if config.model_type in {"bloom", "gpt2", "roberta"}:
        tokenizer = AutoTokenizer.from_pretrained(
            tokenizer_name_or_path,
            cache_dir=model_args.cache_dir,
            use_fast=True,
            revision=model_args.model_revision,
            trust_remote_code=trust_remote_code_tokenizer, # For tokenizer
            use_auth_token=True if model_args.use_auth_token else None,
            add_prefix_space=True,
        )
    else:
        tokenizer = AutoTokenizer.from_pretrained(
            tokenizer_name_or_path,
            cache_dir=model_args.cache_dir,
            use_fast=True,
            revision=model_args.model_revision,
            trust_remote_code=trust_remote_code_tokenizer, # For tokenizer
            use_auth_token=True if model_args.use_auth_token else None,
        )

    model = AutoModelForTokenClassification.from_pretrained(
        model_args.model_name_or_path,
        from_tf=bool(".ckpt" in model_args.model_name_or_path),
        config=config,
        cache_dir=model_args.cache_dir,
        revision=model_args.model_revision,
        trust_remote_code=trust_remote_code_model, # For model
        use_auth_token=True if model_args.use_auth_token else None,
        ignore_mismatched_sizes=model_args.ignore_mismatched_sizes,
    )
```

sgugger commented:
No, one is enough. Users that want more finegrained control can just modify the examples to suit their needs.

datquocnguyen committed:
Merge pull request #41 from huggingface/main

Add up-to-date commits

datquocnguyen committed:
Merge pull request #42 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Add up-to-date commits

datquocnguyen committed:
Merge pull request #43 from huggingface/main

Merge latest commits

datquocnguyen committed:
Merge pull request #44 from datquocnguyen/fast_tokenizers_BARTpho_PhoBERT_BERTweet

Merge latest commits
